File: Spark/Scala/playerWithMaxRunsPerYear.scala

case class Batting(player:String, year:Int, runs:Int);

val batting = sc.textFile("/user/root/pigInputs/batting.csv");

val battingRDD = batting.map(x => {val parts = x.split(","); parts match { case Array(a,b,c) =>  ( c.toInt, Batting(a, b.toInt, c.toInt)) } } );

val runsYear = battingRDD.map(x => (x._2.year, x._2.runs);

val maxRunsByYear = runsYear.reduceByKey( (acc, value) => { if(acc > value) acc else value   });

val joined = maxRunsByYear.join(battingRDD);

val playerMaxRunsPerYear = joined.map(x => (x._1, x._2._2.player, x._2._2.runs);
val sorted = playerMaxRunsPerYear.sortByKey();

File: Spark/Pyspark/playerWithMaxRunsPerYear.py
from pyspark import SparkConf,SparkContext

conf = SparkConf()
conf.setAppName("Max runs per year")

sc = SparkContext(conf=conf)

def parseBatting(rec):
	parts = rec.split(",")
	runs = 0
	if(len(parts[2]) != 0):
		runs = int(parts[2])
	return (parts[0], int(parts[1]), runs )

def topNRuns(values, num):
	runsList = sorted(set(map(lambda x: x[2], values)), reverse = True)[0:num]
	result = []
	for x in values:
		if(x[2] in runsList):
			result.append(x)

	return result

batting = sc.textFile("/user/cloudera/datafiles/common/lahman/batting.csv", False)

battingRec = batting.map(lambda x : parseBatting(x))
battingFilt = battingRec.filter(lambda x : x[2] > 0)
battingGrp = battingFilt.map(lambda x : (x[1], x))

'''
maxRunsPerYear = battingGrp.reduceByKey(lambda x, y: x if(x[2] >= y[2]) else y)

sorted = maxRunsPerYear.sortByKey()

for x in sorted.collect(): print(x)
'''
# Dense Ranking if two or more players have max runs
maxRunsPerYear = battingGrp.groupByKey().flatMapValues(lambda x : topNRuns(x, 1))

sorted = maxRunsPerYear.sortByKey()
for x in sorted.collect(): print(x)
